File: backend/internal/handler/ai.go
package handler

import (
	"context"
	"net/http"

	"ai-writing-assistant/internal/pkg/ai"

	"github.com/gin-gonic/gin"
)

type aiRequest struct {
	Text   string `json:"text"`
	Prompt string `json:"prompt"`
}

type aiResponse struct {
	Result string `json:"result"`
}

// 统一AI请求结构
type unifiedAiRequest struct {
	FunctionType    string `json:"functionType"`
	DocumentSummary string `json:"documentSummary"`
	UserRequirement string `json:"userRequirement"`
	SelectedText    string `json:"selectedText"`
	ContextText     string `json:"contextText"`
	CursorPosition  int    `json:"cursorPosition"`
	ModelName       string `json:"modelName"` // 新增：指定使用的模型
	SessionID       string `json:"sessionID"` // 新增：会话ID，用于多轮对话
	Stream          bool   `json:"stream"`    // 新增：是否启用流式返回
}

// 统一AI响应结构
type unifiedAiResponse struct {
	Result       string `json:"result"`
	FunctionType string `json:"functionType"`
	ModelName    string `json:"modelName"`
}

// 模型切换请求
type switchModelRequest struct {
	ModelName string `json:"modelName"`
}

// 模型切换响应
type switchModelResponse struct {
	Success   bool   `json:"success"`
	Message   string `json:"message"`
	ModelName string `json:"modelName"`
}

// 多轮对话请求
type chatRequest struct {
	Message   string `json:"message"`
	SessionID string `json:"sessionID"`
	ModelName string `json:"modelName"`
}

// 多轮对话响应
type chatResponse struct {
	Result    string `json:"result"`
	ModelName string `json:"modelName"`
}

func registerAIRoutes(g *gin.RouterGroup) {
	svc := ai.NewService()

	// 注册AI提供者
	svc.Register("mock", ai.MockProvider{})
	// svc.Register("tongyi", ai.NewTongyiProvider())
	svc.Register("deepseek", ai.NewDeepSeekProvider())
	// svc.Register("wenxin", ai.NewWenxinProvider())

	// 使用配置管理系统设置默认模型
	config := ai.GetAIConfig()
	svc.Use(config.AI.DefaultModel)

	// 获取可用模型列表
	g.GET("/ai/models", func(c *gin.Context) {
		models := svc.GetAvailableModels()
		c.JSON(http.StatusOK, gin.H{
			"models":  models,
			"current": svc.GetCurrentModel(),
		})
	})

	// 切换AI模型
	g.POST("/ai/switch-model", func(c *gin.Context) {
		var req switchModelRequest
		if err := c.BindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "无效的请求数据"})
			return
		}

		// 检查模型是否可用
		models := svc.GetAvailableModels()
		var targetModel *ai.ModelInfo
		for _, model := range models {
			if model.Name == req.ModelName {
				targetModel = &model
				break
			}
		}

		if targetModel == nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "模型不存在"})
			return
		}

		if !targetModel.IsAvailable {
			c.JSON(http.StatusBadRequest, gin.H{"error": "模型不可用"})
			return
		}

		// 切换模型
		svc.Use(targetModel.Provider)

		c.JSON(http.StatusOK, switchModelResponse{
			Success:   true,
			Message:   "模型切换成功",
			ModelName: req.ModelName,
		})
	})

	// 多轮对话接口
	g.POST("/ai/chat", func(c *gin.Context) {
		var req chatRequest
		if err := c.BindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "无效的请求数据"})
			return
		}

		// 如果指定了模型，先切换
		if req.ModelName != "" && req.ModelName != svc.GetCurrentModel() {
			// 根据模型名称找到对应的provider
			models := svc.GetAvailableModels()
			var targetProvider string
			for _, model := range models {
				if model.Name == req.ModelName {
					targetProvider = model.Provider
					break
				}
			}

			if targetProvider != "" {
				svc.Use(targetProvider)
			}
		}

		// 检查当前provider是否可用
		currentProvider := svc.CurrentProvider()
		if currentProvider == nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "当前AI模型不可用，请先切换模型"})
			return
		}

		// 调用多轮对话
		result, err := currentProvider.Chat(context.Background(), req.Message, req.SessionID)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "AI对话失败: " + err.Error()})
			return
		}

		c.JSON(http.StatusOK, chatResponse{
			Result:    result,
			ModelName: svc.GetCurrentModel(),
		})
	})

	// 统一AI接口（支持流式和非流式）
	g.POST("/ai/unified", func(c *gin.Context) {
		var req unifiedAiRequest
		if err := c.BindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "无效的请求数据"})
			return
		}

		// 如果指定了模型，先切换
		if req.ModelName != "" && req.ModelName != svc.GetCurrentModel() {
			// 根据模型名称找到对应的provider
			models := svc.GetAvailableModels()
			var targetProvider string
			for _, model := range models {
				if model.Name == req.ModelName {
					targetProvider = model.Provider
					break
				}
			}

			if targetProvider != "" {
				svc.Use(targetProvider)
			}
		}

		// 检查当前provider是否可用
		currentProvider := svc.CurrentProvider()
		if currentProvider == nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "当前AI模型不可用，请先切换模型"})
			return
		}

		// 如果启用流式返回
		if req.Stream {
			c.Header("Content-Type", "text/event-stream")
			c.Header("Cache-Control", "no-cache")
			c.Header("Connection", "keep-alive")
			c.Header("Access-Control-Allow-Origin", "*")
			c.Header("Access-Control-Allow-Headers", "Cache-Control")

			// 创建流式响应
			streamWriter := &StreamWriter{c: c}

			var prompt string
			// 根据功能类型构建提示词
			switch req.FunctionType {
			case "continue":
				prompt = buildContinuePrompt(req.DocumentSummary, req.UserRequirement, req.ContextText)
			case "polish":
				prompt = buildPolishPrompt(req.UserRequirement, req.SelectedText)
			case "summarize":
				prompt = buildSummarizePrompt(req.UserRequirement, req.SelectedText)
			case "expand":
				prompt = buildExpandPrompt(req.UserRequirement, req.SelectedText, req.ContextText)
			case "generate":
				prompt = buildGeneratePrompt(req.DocumentSummary, req.UserRequirement, req.ContextText)
			default:
				c.JSON(http.StatusBadRequest, gin.H{"error": "不支持的功能类型"})
				return
			}

			// 调用流式AI接口
			err := currentProvider.CallAIStream(context.Background(), req.FunctionType, prompt, req.SessionID, streamWriter)
			if err != nil {
				streamWriter.WriteError(err.Error())
			}
			return
		}

		// 非流式返回（原有逻辑）
		var result string
		var err error

		// 根据功能类型调用不同的AI服务
		switch req.FunctionType {
		case "continue":
			// 续写：基于选中文本和上下文进行续写
			prompt := buildContinuePrompt(req.DocumentSummary, req.UserRequirement, req.ContextText)
			result, err = currentProvider.ContinueWriting(context.Background(), prompt)
		case "polish":
			// 润色：优化选中文本的表达
			prompt := buildPolishPrompt(req.UserRequirement, req.SelectedText)
			result, err = currentProvider.PolishText(context.Background(), prompt)
		case "summarize":
			// 总结：提取选中文本的核心要点
			prompt := buildSummarizePrompt(req.UserRequirement, req.SelectedText)
			result, err = currentProvider.SummarizeText(context.Background(), prompt)
		case "expand":
			// 扩写：基于选中文本和上下文进行扩写
			prompt := buildExpandPrompt(req.UserRequirement, req.SelectedText, req.ContextText)
			result, err = currentProvider.ContinueWriting(context.Background(), prompt)
		case "generate":
			// 生成：根据用户要求生成新内容
			prompt := buildGeneratePrompt(req.DocumentSummary, req.UserRequirement, req.ContextText)
			result, err = currentProvider.ContinueWriting(context.Background(), prompt)
		default:
			c.JSON(http.StatusBadRequest, gin.H{"error": "不支持的功能类型"})
			return
		}

		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "AI处理失败: " + err.Error()})
			return
		}

		c.JSON(http.StatusOK, unifiedAiResponse{
			Result:       result,
			FunctionType: req.FunctionType,
			ModelName:    svc.GetCurrentModel(),
		})
	})

	// 保持原有接口兼容性
	g.POST("/ai/continue", func(c *gin.Context) {
		var req aiRequest
		if err := c.BindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "无效的请求数据"})
			return
		}

		currentProvider := svc.CurrentProvider()
		if currentProvider == nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "当前AI模型不可用，请先切换模型"})
			return
		}

		result, err := currentProvider.ContinueWriting(context.Background(), req.Prompt)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "AI处理失败: " + err.Error()})
			return
		}
		c.JSON(http.StatusOK, aiResponse{Result: result})
	})

	g.POST("/ai/polish", func(c *gin.Context) {
		var req aiRequest
		if err := c.BindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "无效的请求数据"})
			return
		}

		currentProvider := svc.CurrentProvider()
		if currentProvider == nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "当前AI模型不可用，请先切换模型"})
			return
		}

		result, err := currentProvider.PolishText(context.Background(), req.Text)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "AI处理失败: " + err.Error()})
			return
		}
		c.JSON(http.StatusOK, aiResponse{Result: result})
	})

	g.POST("/ai/summarize", func(c *gin.Context) {
		var req aiRequest
		if err := c.BindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "无效的请求数据"})
			return
		}

		currentProvider := svc.CurrentProvider()
		if currentProvider == nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "当前AI模型不可用，请先切换模型"})
			return
		}

		result, err := currentProvider.SummarizeText(context.Background(), req.Text)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "AI处理失败: " + err.Error()})
			return
		}
		c.JSON(http.StatusOK, aiResponse{Result: result})
	})
}

// 流式写入器
type StreamWriter struct {
	c *gin.Context
}

func (sw *StreamWriter) Write(p []byte) (n int, err error) {
	// 发送SSE格式的数据
	sw.c.Data(http.StatusOK, "text/event-stream", []byte("data: "+string(p)+"\n\n"))
	sw.c.Writer.Flush()
	return len(p), nil
}

func (sw *StreamWriter) WriteError(message string) {
	sw.c.Data(http.StatusOK, "text/event-stream", []byte("data: {\"error\":\""+message+"\"}\n\n"))
	sw.c.Writer.Flush()
}

// 构建续写提示词
func buildContinuePrompt(documentSummary, userRequirement, contextText string) string {
	return "续写要求：" + userRequirement + "\n\n" +
		"上下文：" + contextText + "\n\n" +
		"请直接续写，保持风格一致。"
}

// 构建润色提示词
func buildPolishPrompt(userRequirement, selectedText string) string {
	return "润色要求：" + userRequirement + "\n\n" +
		"原文：" + selectedText + "\n\n" +
		"请直接输出润色后的文本。"
}

// 构建总结提示词
func buildSummarizePrompt(userRequirement, selectedText string) string {
	return "总结要求：" + userRequirement + "\n\n" +
		"原文：" + selectedText + "\n\n" +
		"请直接输出总结内容，用自然段落形式描述。"
}

// 构建扩写提示词
func buildExpandPrompt(userRequirement, selectedText, contextText string) string {
	return "扩写要求：" + userRequirement + "\n\n" +
		"原文：" + selectedText + "\n\n" +
		"上下文：" + contextText + "\n\n" +
		"请直接输出扩写后的内容。"
}

// 构建生成提示词
func buildGeneratePrompt(documentSummary, userRequirement, contextText string) string {
	return "生成要求：" + userRequirement + "\n\n" +
		"文档概要：" + documentSummary + "\n\n" +
		"请直接输出生成的内容。"
}

File: 陈慧源/frontend/src/services/ai.ts
import { getHttpClient } from './api'
import type { AiFunctionType, AiRequestParams, AiResult, TextSelection } from '@/types'

export type AiContinueRequest = { prompt: string }
export type AiPolishRequest = { text: string }
export type AiSummarizeRequest = { text: string }

export type AiResponse = { result: string }

// AI模型信息
export type AiModelInfo = {
	name: string
	displayName: string
	provider: string
	description: string
	maxTokens: number
	isAvailable: boolean
}

// 模型列表响应
export type ModelsResponse = {
	models: AiModelInfo[]
	current: string
}

// 模型切换请求
export type SwitchModelRequest = {
	modelName: string
}

// 模型切换响应
export type SwitchModelResponse = {
	success: boolean
	message: string
	modelName: string
}

// 新的统一AI接口
export type UnifiedAiRequest = {
	functionType: AiFunctionType
	documentSummary: string
	userRequirement: string
	selectedText: string
	contextText: string
	cursorPosition: number
	modelName?: string // 可选：指定使用的模型
	sessionID?: string // 新增：会话ID，用于多轮对话
	stream?: boolean   // 新增：是否启用流式返回
}

export type UnifiedAiResponse = {
	result: string
	functionType: AiFunctionType
	modelName: string
}

// 多轮对话请求
export type ChatRequest = {
	message: string
	sessionID: string
	modelName?: string
}

// 多轮对话响应
export type ChatResponse = {
	result: string
	modelName: string
}

// 流式AI回调函数类型
export type StreamCallback = (chunk: string, isComplete: boolean) => void

// 智能内容选择函数
export function getSmartTextSelection(
  fullText: string,
  cursorPosition: number,
  selectionStart: number,
  selectionEnd: number
): TextSelection {
  const selectedText = fullText.substring(selectionStart, selectionEnd)
  
  // 计算上下文范围（前后各300字符）
  const contextSize = 300
  const contextStart = Math.max(0, selectionStart - contextSize)
  const contextEnd = Math.min(fullText.length, selectionEnd + contextSize)
  
  // 智能调整边界，避免截断句子
  let adjustedContextStart = contextStart
  let adjustedContextEnd = contextEnd
  
  // 向前查找句子边界
  for (let i = selectionStart - 1; i >= contextStart; i--) {
    if (['。', '！', '？', '.', '!', '?', '\n'].includes(fullText[i])) {
      adjustedContextStart = i + 1
      break
    }
  }
  
  // 向后查找句子边界
  for (let i = selectionEnd; i < contextEnd; i++) {
    if (['。', '！', '？', '.', '!', '?', '\n'].includes(fullText[i])) {
      adjustedContextEnd = i + 1
      break
    }
  }
  
  const contextBefore = fullText.substring(adjustedContextStart, selectionStart)
  const contextAfter = fullText.substring(selectionEnd, adjustedContextEnd)
  
  return {
    start: selectionStart,
    end: selectionEnd,
    text: selectedText,
    contextBefore,
    contextAfter
  }
}

// 生成文档摘要
export function generateDocumentSummary(text: string): string {
  if (text.length <= 200) return text
  
  // 简单摘要：取前100字符 + 后100字符
  const prefix = text.substring(0, 100)
  const suffix = text.substring(text.length - 100)
  return `${prefix}...${suffix}`
}

// 生成会话ID
export function generateSessionID(): string {
  return `session_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`
}

// 获取可用模型列表
export async function getAvailableModels(): Promise<ModelsResponse> {
	const res = await getHttpClient().get<ModelsResponse>('/ai/models')
	return res.data
}

// 切换AI模型
export async function switchModel(modelName: string): Promise<SwitchModelResponse> {
	const res = await getHttpClient().post<SwitchModelResponse>('/ai/switch-model', { modelName })
	return res.data
}

// 多轮对话接口
export async function chat(data: ChatRequest): Promise<ChatResponse> {
	const res = await getHttpClient().post<ChatResponse>('/ai/chat', data)
	return res.data
}

// 流式统一AI接口
export async function callUnifiedAIStream(
  data: UnifiedAiRequest, 
  callback: StreamCallback
): Promise<void> {
  try {
    // 在开发环境中，vite代理会将/api转发到http://localhost:8080
    // 在生产环境中，应该设置正确的API基础URL
    const apiUrl = '/api/ai/unified'
    
    // 获取认证token，与api.ts保持一致
    const token = localStorage.getItem('auth-token') || ''
    
    const response = await fetch(apiUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': token, // 直接使用token，不加Bearer前缀，与axios保持一致
        'Accept': 'text/event-stream',
        'Cache-Control': 'no-cache',
        'Connection': 'keep-alive'
      },
      body: JSON.stringify({
        ...data,
        stream: true
      })
    })

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const reader = response.body?.getReader()
    if (!reader) {
      throw new Error('No response body')
    }

    const decoder = new TextDecoder()
    let buffer = ''

    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split('\n')
      buffer = lines.pop() || ''

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          const data = line.slice(6)
          if (data === '[DONE]') {
            callback('', true)
            return
          }

          try {
            const parsed = JSON.parse(data)
            if (parsed.error) {
              throw new Error(parsed.error)
            }
            callback(data, false)
          } catch (e) {
            // 如果不是JSON，直接作为文本内容处理
            callback(data, false)
          }
        }
      }
    }

    callback('', true)
  } catch (error) {
    console.error('流式AI调用失败:', error)
    throw error
  }
}

// 统一AI接口（非流式）
export async function callUnifiedAI(data: UnifiedAiRequest): Promise<UnifiedAiResponse> {
	const res = await getHttpClient().post<UnifiedAiResponse>('/ai/unified', data)
	return res.data
}

// 原有接口保持不变
export async function continueWriting(data: AiContinueRequest) {
  const res = await getHttpClient().post<AiResponse>('/ai/continue', data)
  return res.data
}

export async function polishText(data: AiPolishRequest) {
  const res = await getHttpClient().post<AiResponse>('/ai/polish', data)
  return res.data
}

export async function summarizeText(data: AiSummarizeRequest) {
  const res = await getHttpClient().post<AiResponse>('/ai/summarize', data)
  return res.data
}

